### PR TITLE
PDF contrast: incorrect set by a gesture

### PR DESCRIPTION
**Pull request by hius07**
Closes https://github.com/koreader/koreader/issues/10795.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10798)

### CODE REVIEW

**Comment by hius07** [comment]
(1) The item is called "Contrast" in the bottom menu and in the Taps and gestures menu. Notification shows "Font gamma set to" message. Should it be changed to "Contrast"?

(2) Do we need the more_options `⋮` item in the bottom menu next to the progress bar?

(3) If not (2), do we need the SpinWidget in the Taps and gestures actions menu?
Or the fixed set of values `{0.8, 1.0, 1.5, 2.0, 4.0, 6.0, 10.0, 50.0}` is enough?

File: frontend/apps/reader/modules/readerkoptlistener.lua [hunk 1]
```diff
@@ -21,10 +21,7 @@ function ReaderKoptListener:onReadSettings(config)
     normal_zoom_mode = ReaderZooming.zoom_mode_label[normal_zoom_mode] and normal_zoom_mode or ReaderZooming.DEFAULT_ZOOM_MODE
     self.normal_zoom_mode = normal_zoom_mode
     self:setZoomMode(normal_zoom_mode)
-    self.document.configurable.contrast = config:readSetting("kopt_contrast")
-                                       or G_reader_settings:readSetting("kopt_contrast")
-                                       or 1.0
-    self.ui:handleEvent(Event:new("GammaUpdate", 1/self.document.configurable.contrast))
+    self.ui:handleEvent(Event:new("GammaUpdate", self.document.configurable.contrast))
```
Comment (yparitcher) [inline frontend/apps/reader/modules/readerkoptlistener.lua:24]: Maybe move this to `ReaderView` where the rest of it is?

**Review by yparitcher — APPROVED** [review]
1. Yes

**Comment by poire-z** [comment]
> Notification shows "Font gamma set to" message. Should it be changed to "Contrast"?

Is that just about PDF ?
In CRE, this contrast applies only to text/font - not to images. So, mentionning "font" may be fine.
I guess "font gamma" is quite strange, dunno how many people know what "gamma" means (i don't :) and I can't even find anything via Google...)

And I think this contrast/gamma in CRE should not apply to images. I sometimes miss another setting that would apply to images that could make grey text in images more readable :) I'm usually annoyed when I need more contrast in PDF for text, and it make all images nearly black... but I guess we can't distinguish text from images in PDF.

**Comment by hius07** [comment]
Yes, this PR is about pdf only.
https://github.com/koreader/koreader/blob/ac143ca9d69456c56fc913c5d6d7fe847ddc0620/frontend/apps/reader/modules/readerview.lua#L997
cre shows the same:
https://github.com/koreader/koreader/blob/ac143ca9d69456c56fc913c5d6d7fe847ddc0620/frontend/apps/reader/modules/readerfont.lua#L350

File: frontend/apps/reader/modules/readerkoptlistener.lua (hunk 1 above)
Comment (hius07) [reply frontend/apps/reader/modules/readerkoptlistener.lua:24]: We can't, the "configurable" settings are loaded in ReaderConfig that is registered in RederUI after ReaderView.

Comment (hius07) [reply frontend/apps/reader/modules/readerkoptlistener.lua:24]: So ReaderView handles "ReadSettings" event before ReaderConfig.

**Comment by hius07** [comment]
> (2) Do we need the more_options `⋮` item in the bottom menu next to the progress bar?

For cre we have the vice versa inconsistency: the bottom menu has more_options `⋮` item, but Taps and gestures actions menu shows Contrast preset values, not the SpinWidget.

File: frontend/ui/data/koptoptions.lua [hunk 2]
```diff
@@ -481,13 +481,10 @@ Some of the other settings are only available when reflow mode is enabled.]]),
                 name = "contrast",
                 name_text = _("Contrast"),
                 buttonprogress = true,
-                -- See https://github.com/koreader/koreader/issues/1299#issuecomment-65183895
-                -- For pdf reflowing mode (kopt_contrast):
```
Comment (NiLuJe) [inline frontend/ui/data/koptoptions.lua:485]: I would very much keep those comments around ;).

Comment (NiLuJe) [reply frontend/ui/data/koptoptions.lua:485]: With an addendum that the div was moved to KoptInterface ;).

**Comment by hius07** [comment]
https://github.com/koreader/koreader/blob/0db7509fe244c7a5ad6710a8aea8b9d22143f9c9/frontend/ui/data/koptoptions.lua#L375
What is optimized by calling a function instead of `return configurable.text_wrap == 1`?

**Comment by poire-z** [comment]
>  What is optimized by calling a function instead of return configurable.text_wrap == 1?

Dunno, maybe nothing - it seems a remnant from some old PR where the author decided to have other such functions to maybe make things more readable? #1455.

**Comment by Frenzie** [comment]
> What is optimized by calling a function instead of return configurable.text_wrap == 1?

I think the answer to most such questions would be clarity and/or maintainability? :-)

**Comment by hius07** [comment]
So, let's keep it?
Is it a clarity - to change `==` with a word "equal"?
Easy to maintain? If some day the reflow mode is detected in other way, it will definitely require to change all the calls, not only the function itself.

**Comment by poire-z** [comment]
Do as you prefer :)
(My habit is to not change what I don't need to - but as you have become the boss of the bottom menu hell, you're free to do as you like :))

**Comment by Frenzie** [comment]
> Easy to maintain?

I basically mean greppability. It looks to me like `optionsutil.enableIfEquals` has a very high greppability factor, which normally counts as a positive in my book.

NB I'm not saying I'd have written it the same way or that I'm opposed to changing it; I barely know what the context is here. It's been at least a couple of years since I last dove into this code. I'm just speaking generically about code/maintenance patterns.

**Comment by hius07** [comment]
Update: make pdf Contrast consistent to cre Contrast (⋮ next to the progress bar in the bottom menu; no SpinWidget in Taps and gestures menu, just eight preset values).

**Comment by hius07** [comment]
I believe the bug should be fixed.

**Comment by poire-z** [comment]
Looks ok to me. @yparitcher : ok for you ?
Please juste update `CURRENT_MIGRATION_DATE = 20230810` to today :)

**Comment by yparitcher** [comment]
Does this work if the user has a gesture/profile saved with the old values?
Also the migration only fixes the global setting, what if i open a old document?
Not at my computer and I don't remember if we store the value or the arg or label.

**Comment by hius07** [comment]
Old setting will become new setting
https://github.com/koreader/koreader/blob/5f336a76870b8477268255faf5900ddf822f9513/frontend/apps/reader/modules/readerview.lua#L863-L866
And old calls
https://github.com/koreader/koreader/blob/ed2ea6803f4b9be6b065f43590229dea4fe51426/frontend/apps/reader/modules/readerkoptlistener.lua#L27
https://github.com/koreader/koreader/blob/ed2ea6803f4b9be6b065f43590229dea4fe51426/frontend/document/koptinterface.lua#L147
will become new calls
https://github.com/koreader/koreader/blob/5f336a76870b8477268255faf5900ddf822f9513/frontend/apps/reader/modules/readerkoptlistener.lua#L24
https://github.com/koreader/koreader/blob/5f336a76870b8477268255faf5900ddf822f9513/frontend/document/koptinterface.lua#L148
Following https://github.com/koreader/koreader/issues/1299#issuecomment-65183895, instead of saving 2 values for non-reflow and reflow modes (gamma and 1/gamma), we now save only one value, and in reflow mode use 1/value as argument to `setContrast`.

**Comment by yparitcher** [comment]
Does the config value need to be updated `config:saveSetting("kopt_contrast", 1/ config:readSetting("gamma"))`?

Also what about gestures that store call the gamma event directly?

**Comment by hius07** [comment]
> Does the config value need to be updated

No, because option values are changed too.
Before:
https://github.com/koreader/koreader/blob/ed2ea6803f4b9be6b065f43590229dea4fe51426/frontend/ui/data/koptoptions.lua#L486
After:
https://github.com/koreader/koreader/blob/5f336a76870b8477268255faf5900ddf822f9513/frontend/ui/data/koptoptions.lua#L486

**Review by yparitcher — APPROVED** [review]
I see now.
Sorry for being slow today :)

**Pull request merged**